// @graphql-codegen/typescript-react-apollo/dist/commonjs/visitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const visitor_plugin_common_1 = require("@graphql-codegen/visitor-plugin-common");
const autoBind = require("auto-bind");
const plugin_helpers_1 = require("@graphql-codegen/plugin-helpers");
const change_case_1 = require("change-case");
class ReactApolloVisitor extends visitor_plugin_common_1.ClientSideBaseVisitor {
    constructor(fragments, rawConfig) {
        super(fragments, rawConfig, {
            withHOC: visitor_plugin_common_1.getConfigValue(rawConfig.withHOC, true),
            withComponent: visitor_plugin_common_1.getConfigValue(rawConfig.withComponent, true),
            withHooks: visitor_plugin_common_1.getConfigValue(rawConfig.withHooks, false),
            hooksImportFrom: visitor_plugin_common_1.getConfigValue(rawConfig.hooksImportFrom, 'react-apollo-hooks'),
        });
        autoBind(this);
    }
    getImports() {
        const baseImports = super.getImports();
        const imports = [];
        if (this.config.withComponent) {
            imports.push(`import * as React from 'react';`);
        }
        if (this.config.withComponent || this.config.withHOC) {
            imports.push(`import * as ReactApollo from 'react-apollo';`);
        }
        if (this.config.withHooks) {
            imports.push(`import * as ReactApolloHooks from '${typeof this.config.hooksImportFrom === 'string' ? this.config.hooksImportFrom : 'react-apollo-hooks'}';`);
        }
        return [baseImports, ...imports].join('\n');
    }
    _buildHocProps(operationName, operationType) {
        const typeVariableName = this.convertName(operationName + plugin_helpers_1.toPascalCase(operationType));
        const variablesVarName = this.convertName(operationName + plugin_helpers_1.toPascalCase(operationType) + 'Variables');
        const argType = operationType === 'mutation' ? 'MutateProps' : 'DataProps';
        return `Partial<ReactApollo.${argType}<${typeVariableName}, ${variablesVarName}>>`;
    }
    _buildOperationHoc(node, documentVariableName, operationResultType, operationVariablesTypes) {
        const operationName = this.convertName(node.name.value);
        const propsTypeName = operationName + 'Props';
        const propsVar = `export type ${propsTypeName}<TChildProps = {}> = ${this._buildHocProps(node.name.value, node.operation)} & TChildProps;`;
        const mutationFn = node.operation === 'mutation' ? `export type ${this.convertName(node.name.value + 'MutationFn')} = ReactApollo.MutationFn<${operationResultType}, ${operationVariablesTypes}>;` : null;
        const hocString = `export function with${operationName}<TProps, TChildProps = {}>(operationOptions: ReactApollo.OperationOption<
  TProps, 
  ${operationResultType},
  ${operationVariablesTypes},
  ${propsTypeName}<TChildProps>> | undefined) {
    return ReactApollo.with${change_case_1.titleCase(node.operation)}<TProps, ${operationResultType}, ${operationVariablesTypes}, ${propsTypeName}<TChildProps>>(${documentVariableName}, operationOptions);
};`;
        return [propsVar, mutationFn, hocString].filter(a => a).join('\n');
    }
    _buildComponent(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        const componentName = this.convertName(node.name.value + 'Component');
        return `
export class ${componentName} extends React.Component<Partial<ReactApollo.${operationType}Props<${operationResultType}, ${operationVariablesTypes}>>> {
  render() {
      return (
          <ReactApollo.${operationType}<${operationResultType}, ${operationVariablesTypes}> ${node.operation}={${documentVariableName}} {...(this as any)['props'] as any} />
      );
  }
}`;
    }
    _buildHooks(node, operationType, documentVariableName, operationResultType, operationVariablesTypes) {
        return `
export function use${operationResultType}(baseOptions?: ReactApolloHooks.${operationType}HookOptions<${node.operation !== 'query' ? `${operationResultType}, ` : ''}${operationVariablesTypes}>) {
  return ReactApolloHooks.use${operationType}<${operationResultType}, ${operationVariablesTypes}>(${documentVariableName}, baseOptions);
};`;
    }
    buildOperation(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        const component = this.config.withComponent ? this._buildComponent(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) : null;
        const hoc = this.config.withHOC ? this._buildOperationHoc(node, documentVariableName, operationResultType, operationVariablesTypes) : null;
        const hooks = this.config.withHooks ? this._buildHooks(node, operationType, documentVariableName, operationResultType, operationVariablesTypes) : null;
        return [component, hoc, hooks].filter(a => a).join('\n');
    }
}
exports.ReactApolloVisitor = ReactApolloVisitor;
