// @graphql-codegen/typescript-react-apollo/dist/visitor.js
import { ClientSideBaseVisitor, getConfigValue } from 'graphql-codegen-visitor-plugin-common';
import autoBind from 'auto-bind';
import { toPascalCase } from 'graphql-codegen-plugin-helpers';
export class ReactApolloVisitor extends ClientSideBaseVisitor {
    constructor(fragments, rawConfig) {
        super(fragments, rawConfig, {
            withHOC: getConfigValue(rawConfig.withHOC, true),
            withComponent: getConfigValue(rawConfig.withComponent, true),
            withHooks: getConfigValue(rawConfig.withHooks, false),
            hooksImportFrom: getConfigValue(rawConfig.hooksImportFrom, 'react-apollo-hooks')
        });
        autoBind(this);
    }
    getImports() {
        const baseImports = super.getImports();
        const imports = [];
        if (this.config.withComponent) {
            imports.push(`import * as React from 'react';`);
        }
        if (this.config.withComponent || this.config.withHOC) {
            imports.push(`import * as ReactApollo from 'react-apollo';`);
        }
        if (this.config.withHooks) {
            imports.push(`import * as ReactApolloHooks from '${typeof this.config.hooksImportFrom === 'string' ? this.config.hooksImportFrom : 'react-apollo-hooks'}';`);
        }
        return [baseImports, ...imports].join('\n');
    }
    _buildHocProps(operationName, operationType) {
        const typeVariableName = this.convertName(operationName + toPascalCase(operationType));
        const variablesVarName = this.convertName(operationName + toPascalCase(operationType) + 'Variables');
        const argType = operationType === 'mutation' ? 'MutateProps' : 'DataProps';
        return `Partial<ReactApollo.${argType}<${typeVariableName}, ${variablesVarName}>>`;
    }
    _buildOperationHoc(node, documentVariableName, operationResultType, operationVariablesTypes) {
        const operationName = this.convertName(node.name.value);
        const propsTypeName = operationName + 'Props';
        const propsVar = `export type ${propsTypeName}<TChildProps = any> = ${this._buildHocProps(node.name.value, node.operation)} & TChildProps;`;
        const mutationFn = node.operation === 'mutation'
            ? `export type ${this.convertName(node.name.value + 'MutationFn')} = ReactApollo.MutationFn<${operationResultType}, ${operationVariablesTypes}>;`
            : null;
        const hocString = `export function ${operationName}HOC<TProps, TChildProps = any>(operationOptions: ReactApollo.OperationOption<
  TProps, 
  ${operationResultType},
  ${operationVariablesTypes},
  ${propsTypeName}<TChildProps>> | undefined) {
    return ReactApollo.graphql<TProps, ${operationResultType}, ${operationVariablesTypes}, ${propsTypeName}<TChildProps>>(${documentVariableName}, operationOptions);
};`;
        return [propsVar, mutationFn, hocString].filter(a => a).join('\n');
    }
    _buildComponent(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        const componentName = this.convertName(node.name.value + 'Component');
        return `
export class ${componentName} extends React.Component<Partial<ReactApollo.${operationType}Props<${operationResultType}, ${operationVariablesTypes}>>> {
  render() {
      return (
          <ReactApollo.${operationType}<${operationResultType}, ${operationVariablesTypes}> ${node.operation}={${documentVariableName}} {...(this as any)['props'] as any} />
      );
  }
}`;
    }
    _buildHooks(node, operationType, documentVariableName, operationResultType, operationVariablesTypes) {
        return `
export function use${operationResultType}(baseOptions?: ReactApolloHooks.${operationType}HookOptions<${node.operation !== 'query' ? `${operationResultType}, ` : ''}${operationVariablesTypes}>) {
  return ReactApolloHooks.use${operationType}<${operationResultType}, ${operationVariablesTypes}>(${documentVariableName}, baseOptions);
};`;
    }
    buildOperation(node, documentVariableName, operationType, operationResultType, operationVariablesTypes) {
        const component = this.config.withComponent
            ? this._buildComponent(node, documentVariableName, operationType, operationResultType, operationVariablesTypes)
            : null;
        const hoc = this.config.withHOC
            ? this._buildOperationHoc(node, documentVariableName, operationResultType, operationVariablesTypes)
            : null;
        const hooks = this.config.withHooks
            ? this._buildHooks(node, operationType, documentVariableName, operationResultType, operationVariablesTypes)
            : null;
        return [component, hoc, hooks].filter(a => a).join('\n');
    }
}
